### PR TITLE
ghc@8.6: use intree gmp for Linux

### PR DESCRIPTION
**Pull request by iMichka**
- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [ ] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

**Comment by carlocab** [comment]
You forgot
```ruby
uses_from_macos "m4" => :build
```
Also in your other `ghc` PR.

File: Formula/ghc@8.6.rb [hunk 1]
```diff
@@ -20,14 +20,22 @@ class GhcAT86 < Formula
...
+    resource "gmp" do
+      url "https://ftp.gnu.org/gnu/gmp/gmp-6.1.2.tar.xz"
+      mirror "https://gmplib.org/download/gmp/gmp-6.1.2.tar.xz"
+      mirror "https://ftpmirror.gnu.org/gmp/gmp-6.1.2.tar.xz"
+      sha256 "87b565e89a9a684fe4ebeeddb8399dce2599f9c9049854ca8c0dfbdea0e21912"
+    end
+  end
+
+  on_linux do
+    depends_on "gmp" => :build
```
Comment (cho-m) [inline Formula/ghc@8.6.rb:38]: Does the `linuxbrew-core` formula actually use `gmp` formula or does it statically link the same resource as macOS? The upstream formula is still building/installing a static `gmp` in the `ghc@8.6` formula https://github.com/Homebrew/linuxbrew-core/blob/master/Formula/ghc%408.6.rb#L69-L81

Current linkage failure with `gmp` formula:
```
==> brew linkage --test ghc@8.6
==> FAILED
Unwanted system libraries:
  /usr/lib/x86_64-linux-gnu/libgmp.so.10
```

Comment (danielnachun) [reply Formula/ghc@8.6.rb:38]: `gmp` is a dependency here for the boostrap binaries, which apparently are not statically linked to `libgmp`.

**Comment by danielnachun** [comment]
From a local build:
```
config.log:  $ ./configure --prefix=/home/linuxbrew/.linuxbrew/Cellar/ghc@8.6/8.6.5_2 --enable-numa=no --with-gmp-includes=/home/linuxbrew/.linuxbrew/opt/gmp/include --with-gmp-libraries=/home/linuxbrew/.linuxbrew/opt/gmp/lib --with-intree-gmp
config.log:configure:16814: WARNING: unrecognized options: --enable-numa, --with-intree-gmp
```
It looks like they didn't provide this option in `ghc@8.6`. Instead, I found this in the `mk/build.mk.sample` file:
```
# Uncomment the following to force `integer-gmp` to use the in-tree GMP 6.1.2
# (other sometimes useful configure-options: `--with-gmp-{includes,libraries}`)
#libraries/integer-gmp_CONFIGURE_OPTS += --configure-option=--with-intree-gmp
```
I will try adding this to the `build.mk` we already have and see if that is sufficient to make it statically link to `gmp`.

**Comment by danielnachun** [comment]
I tried this locally and the change I proposed worked.  I am going to push it to this repo and we'll see if everything passes the test!

**Comment by danielnachun** [comment]
The fix worked!  This is now ready to review.

File: Formula/ghc@8.6.rb [hunk 2]
```diff
@@ -52,31 +60,46 @@ def install
...
+      # is mandatory or else you'll get "illegal text relocs" errors.
+      resource("gmp").stage do
+        system "./configure", "--prefix=#{gmp}", "--with-pic", "--disable-shared",
+                              "--build=#{Hardware.oldest_cpu}-apple-darwin#{OS.kernel_version.major}"
+        system "make"
+        system "make", "install"
+      end
+
+      args = ["--with-gmp-includes=#{gmp}/include",
+              "--with-gmp-libraries=#{gmp}/lib"]
```
Comment (danielnachun) [inline Formula/ghc@8.6.rb:79]: I'm wondering if we even need this any of our GHC formulae.  We might want to check in a follow up PR if the `build.mk` change I made and `--with-intree-gmp` in the newer GHC formulae is sufficient to statically link `gmp` on macOS as well.  If so, we'd be able to reduce the difference between macOS and Linux in the GHC formulae to just using different bootstrap binaries.

**Review by danielnachun — APPROVED** [review]

File: Formula/ghc@8.6.rb (hunk 2 above)
Comment (carlocab) [reply Formula/ghc@8.6.rb:79]: Yea, I suspect we should be able to do the same thing for macOS that we currently do on Linux. That's been on my to-do list for a while, but haven't found the time to do it yet.

**Pull request merged**